File: Leetcode/1239/1239.cpp
class Solution {
public:
    int ans, n;
    void setBit(int &s, int j) {
        s |= (1 << j);
    }
    bool isOn(int s, int j) {
        return (s & (1 << j));
    }
    void dfs(const vector<string> &arr, int i, int length, int status) {
        ans = max(ans,length);
        bool noDup;
        int statusInCurrStr;
        for (int j = i+1; j < n; ++j) {
            noDup = true;
            for (const char &ch : arr[j]) {
                if (isOn(status, ch - 'a')) {
                    noDup = false;
                    break;
                }
            }
            if (noDup) {
                statusInCurrStr = 0;
                for (const char &ch : arr[j]) {
                    if (isOn(statusInCurrStr, ch - 'a')) {
                        noDup = false;
                        break;
                    }
                    setBit(statusInCurrStr, ch - 'a');
                }
            }
            if (noDup) {
                int newStatus = status;
                for (const char &ch : arr[j]) {
                    setBit(newStatus, ch - 'a');
                }
                dfs(arr,j,length+(int)arr[j].size(),newStatus);
            }
        }
    }
    int maxLength(vector<string>& arr) {
        ans = 0;
        n = (int)arr.size();
        bool noDup;
        int status;
        for (int i = 0; i < n; ++i) {
            status = 0;
            noDup = true;
            for (const char &ch : arr[i]) {
                if (isOn(status, ch - 'a')) {
                    noDup = false;
                    break;
                }
                setBit(status, ch - 'a');
            }
            if (noDup) {
                dfs(arr,i,(int)arr[i].size(),status);
            }
        }
        return ans;
    }
};

File: Leetcode/1239/1239.go
const MAXN int = 26

var (
	bs []bool
	freq [MAXN]int
)

func f(n int, d int) {
	for i := 0; i < d; i++ {
		bs[i] = false
	}
	index := d-1
	for n > 0 {
		if (n & 1) != 0 {
			bs[index] = true
		}
		n >>= 1
		index--
	}
}

func maxLength(arr []string) int {
	ans := 0
	total := 1<<len(arr)
	bs = nil
	bs = make([]bool, len(arr))
	var temp int
	var flag bool
	for i := 1; i < total; i++ {
		f(i, len(arr))
		for j := 0; j < MAXN; j++ {
			freq[j] = 0
		}
		flag = true
		temp = 0
		for j := 0; j < len(arr); j++ {
			if bs[j] {
				temp += len(arr[j])
				for k := 0; k < len(arr[j]); k++ {
					freq[arr[j][k] - 'a']++
					if freq[arr[j][k] - 'a'] > 1 {
						flag = false
						break
					}
				}
			}
			if !flag {
				break
			}
		}
		if flag && temp > ans {
			ans = temp
		}
	}
	return ans
}